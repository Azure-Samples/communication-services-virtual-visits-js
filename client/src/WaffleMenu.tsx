// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import {
  IconButton,
  INavLinkGroup,
  IPanelProps,
  Nav,
  Panel,
  PanelType,
  ThemeContext,
  createTheme
} from '@fluentui/react';
import { Theme } from '@fluentui/theme';
import { useState } from 'react';
import {
  headerMenuIconStyles,
  panelMenuIconStyles,
  wafflePanelNavigationStyles,
  wafflePanelStyles
} from './styles/Header.styles';

interface WaffleMenuProps {
  parentid: string;
}

export const WaffleMenu = (props: WaffleMenuProps): JSX.Element => {
  const { parentid } = props;

  const [isOpen, setIsOpen] = useState<boolean>(false);

  const _openPanel = (): void => {
    setIsOpen(true);
  };

  const _dismissPanel = (): void => {
    setIsOpen(false);
  };

  const _onRenderPanelMenu = (props: IPanelProps | undefined, theme: Theme | undefined): JSX.Element => {
    return (
      <>
        <IconButton
          id="waffle-menu-panel"
          iconProps={{ iconName: 'Waffle' }}
          aria-label="Menu Button"
          styles={theme ? panelMenuIconStyles(theme) : undefined}
          onClick={() => _dismissPanel()}
        />
      </>
    );
  };

  return (
    <ThemeContext.Consumer>
      {(theme: Theme | undefined) => {
        if (theme === undefined) {
          theme = createTheme();
        }
        return (
          <div>
            <IconButton
              id="waffle-menu"
              iconProps={{ iconName: 'Waffle' }}
              aria-label="Menu Button"
              styles={headerMenuIconStyles(theme)}
              onClick={() => _openPanel()}
            />
            <Panel
              styles={wafflePanelStyles(theme)}
              isLightDismiss={true}
              isOpen={isOpen}
              hasCloseButton={false}
              onDismiss={() => _dismissPanel()}
              type={PanelType.smallFixedNear}
              layerProps={{ hostId: parentid }}
              onRenderNavigationContent={(props) => _onRenderPanelMenu(props, theme)}
            >
              <WaffleNavigation />
            </Panel>
          </div>
        );
      }}
    </ThemeContext.Consumer>
  );
};

const navLinkGroups: INavLinkGroup[] = [
  {
    links: [
      {
        name: 'Home',
        url: './',
        key: 'home'
      },
      {
        name: 'Book',
        url: './book',
        key: 'book'
      },
      {
        name: 'Visit',
        url: './visit',
        key: 'visit'
      }
    ]
  }
];

export const WaffleNavigation = (): JSX.Element => {
  return <Nav ariaLabel="Navigate to expected page" styles={wafflePanelNavigationStyles} groups={navLinkGroups} />;
};
