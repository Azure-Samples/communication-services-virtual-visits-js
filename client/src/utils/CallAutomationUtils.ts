// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { CommunicationUserIdentifier } from '@azure/communication-common';
import { CallAdapterState, CommonCallAdapter } from '@azure/communication-react';

export type SummarizeResult = {
  recap: string;
  chapters: {
    chapterTitle: string;
    narrative: string;
  }[];
};

export type CallTranscription = {
  text: string;
  confidence: number;
  participant: CommunicationUserIdentifier;
  resultState: 'intermediate' | 'final';
}[];

/**
 * Fetches the transcription for the given server call ID.
 * @param serverCallId - The server call ID to fetch the transcription for.
 * @returns - the CallTranscription object containing the transcription data.
 */
export const fetchTranscript = async (serverCallId: string): Promise<CallTranscription> => {
  const response = await fetch(`/api/fetchTranscript`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcript: ' + response.statusText);
  }

  return ((await response.json()) as { transcript: CallTranscription }).transcript;
};

/**
 * fetches the transcription status for the given server call ID. This is called when
 * the user first joins the call to check if transcription is already started.
 * @param serverCallId - The server call ID to fetch the transcription status for.
 */
export const fetchTranscriptionStatus = async (serverCallId: string): Promise<void> => {
  const response = await fetch(`/api/fetchTranscriptionState`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcription status: ' + response.statusText);
    return;
  }
  return;
};

/**
 * Starts the transcription for the call.
 * @param serverCallId - The server call ID to start transcription for.
 * @param transcriptionOptions - The options for transcription, including locale.
 * @returns - True if the transcription was started successfully, false otherwise.
 */
export const startTranscription = async (
  serverCallId: string,
  transcriptionOptions?: {
    locale?: string;
  }
): Promise<boolean> => {
  console.log('Starting transcription for call:', serverCallId);
  const response = await fetch('/api/startTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId,
      options: transcriptionOptions
    })
  });
  if (!response.ok) {
    console.error('Failed to start transcription:', response);
    return false;
  }
  console.log('Started transcription:', transcriptionOptions);
  return true;
};

/**
 * Stops the transcription for the call.
 * @param serverCallId - The server call ID to stop transcription for.
 * @returns {boolean} - True if the transcription was stopped successfully, false otherwise.
 */
export const stopTranscription = async (serverCallId: string): Promise<boolean> => {
  console.log('Stopping transcription for call:', serverCallId);

  const response = await fetch('/api/stopTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to stop transcription:', response);
    return false;
  }
  console.log('Stopped transcription');
  return true;
};

/**
 * Connect the Call Automation service to the call.
 * @param callAdaterState - The call adapter state to check for the call.
 * @returns
 */
export const connectToCallAutomation = async (callAdaterState: CallAdapterState): Promise<boolean> => {
  if (callAdaterState.call?.info !== undefined && callAdaterState.call?.state === 'Connected') {
    const serverCallID = await callAdaterState.call.info.getServerCallId();
    const response = await fetch('/api/connectRoomsCall', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        serverCallId: serverCallID
      })
    });
    if (!response.ok) {
      throw new Error('Failed to start call with transcription');
    }
    return true;
  }
  return false;
};

/**
 * Fetch the call summary from the server.
 * @param adapter - The call adapter to get the call ID from.
 * @param locale - locale code for the language to be used in the summary.
 * @returns summary result if there is a summary available.
 */
export const getCallSummaryFromServer = async (
  adapter: CommonCallAdapter,
  locale: LocaleCode
): Promise<SummarizeResult> => {
  console.log('Getting summary from server...');
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const serverCallId = await adapter.getState().call?.info?.getServerCallId();
    if (!serverCallId) {
      console.error('Call ID not found');
      throw new Error('Call ID not found');
    }

    const response = await fetch('/api/summarizeTranscript', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ serverCallId: serverCallId, locale: locale })
    });

    if (!response.ok) {
      console.error('Response Failed: ', response.statusText);
      throw new Error('Summarization request failed');
    }

    const result = await response.json();
    console.log('Summary result:', result);
    return result as SummarizeResult;
  } catch (error) {
    console.error('Error fetching summary:', error);
    throw error;
  }
};

/**
 * Update the participants in the call record for transcription.
 * @param participant - The participant to add to the call record.
 */
export const sendParticipantInfoToServer = async (
  participant: { userId: string; displayName: string },
  serverCallId?: string
): Promise<void> => {
  if (!serverCallId) {
    console.error('Server call ID not found');
    return;
  }
  const response = await fetch('/api/updateParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId,
      participant: participant
    })
  });
  if (!response.ok) {
    console.error('Failed to update participants:', response);
  }
};

/**
 * Fetches the participants in the call. This is needed for when we are computing the transcript for download.
 * @param serverCallId - The server call ID to fetch the participants for.
 * @returns display name and user ID of the participants in the call.
 */
export const fetchParticipants = async (serverCallId: string): Promise<{ userId: string; displayName: string }[]> => {
  const response = await fetch('/api/fetchParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to fetch participants:', response);
    return [];
  }
  return ((await response.json()) as { participants: { userId: string; displayName: string }[] }).participants;
};

/**
 * Language codes used in the localization of the components.
 */
export type LocaleCode =
  | 'ar-AE'
  | 'ar-SA'
  | 'da-DK'
  | 'de-DE'
  | 'en-AU'
  | 'en-CA'
  | 'en-GB'
  | 'en-IN'
  | 'en-NZ'
  | 'en-US'
  | 'es-ES'
  | 'es-MX'
  | 'fi-FI'
  | 'fr-CA'
  | 'fr-FR'
  | 'hi-IN'
  | 'it-IT'
  | 'ja-JP'
  | 'ko-KR'
  | 'nb-NO'
  | 'nl-BE'
  | 'nl-NL'
  | 'pl-PL'
  | 'pt-BR'
  | 'ru-RU'
  | 'sv-SE'
  | 'zh-CN'
  | 'zh-HK'
  | 'cs-CZ'
  | 'pt-PT'
  | 'tr-TR'
  | 'vi-VN'
  | 'th-TH'
  | 'he-IL'
  | 'cy-GB'
  | 'uk-UA'
  | 'el-GR'
  | 'hu-HU'
  | 'ro-RO'
  | 'sk-SK'
  | 'zh-TW';

/**
 * Mapping of locale codes to their display names.
 */
export const localeDisplayNames: Record<LocaleCode, string> = {
  'ar-AE': 'Arabic - U.A.E.',
  'ar-SA': 'Arabic - Saudi Arabia',
  'da-DK': 'Danish',
  'de-DE': 'German - Germany',
  'en-AU': 'English - Australia',
  'en-CA': 'English - Canada',
  'en-GB': 'English - United Kingdom',
  'en-IN': 'English - India',
  'en-NZ': 'English - New Zealand',
  'en-US': 'English - United States',
  'es-ES': 'Spanish - Spain (Modern Sort)',
  'es-MX': 'Spanish - Mexico',
  'fi-FI': 'Finnish',
  'fr-CA': 'French - Canada',
  'fr-FR': 'French - France',
  'hi-IN': 'Hindi',
  'it-IT': 'Italian - Italy',
  'ja-JP': 'Japanese',
  'ko-KR': 'Korean',
  'nb-NO': 'Norwegian (Bokm√•l)',
  'nl-BE': 'Dutch - Belgium',
  'nl-NL': 'Dutch - Netherlands',
  'pl-PL': 'Polish',
  'pt-BR': 'Portuguese - Brazil',
  'ru-RU': 'Russian',
  'sv-SE': 'Swedish',
  'zh-CN': "Chinese - People's Republic of China",
  'zh-HK': 'Chinese - Hong Kong SAR',
  'cs-CZ': 'Czech',
  'pt-PT': 'Portuguese - Portugal',
  'tr-TR': 'Turkish',
  'vi-VN': 'Vietnamese',
  'th-TH': 'Thai',
  'he-IL': 'Hebrew',
  'cy-GB': 'Welsh',
  'uk-UA': 'Ukrainian',
  'el-GR': 'Greek',
  'hu-HU': 'Hungarian',
  'ro-RO': 'Romanian',
  'sk-SK': 'Slovak',
  'zh-TW': 'Chinese - Taiwan'
};
