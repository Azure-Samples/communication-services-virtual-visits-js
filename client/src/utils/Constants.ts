// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export const BOOKINGS_README_URL = 'https://aka.ms/virtual-appointments-sample-bookings';
export const BOOKINGS_SPECIMEN_URL =
  'https://microsoftbookings.azurewebsites.net/?organization=financialservices&UICulture=en-US';
