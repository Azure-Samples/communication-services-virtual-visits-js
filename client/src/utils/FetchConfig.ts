// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { AppConfigModel } from '../models/ConfigModel';
import { generateTheme } from './ThemeGenerator';
import createError from 'http-errors';

export async function fetchConfig(): Promise<AppConfigModel | undefined> {
  const response = await fetch('/api/config', { method: 'GET' });
  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const config = JSON.parse(responseContent);
  if (!config) {
    throw new Error(`Received empty config response.`);
  }

  const appConfigModel: AppConfigModel = {
    ...config,
    theme: generateTheme(config.colorPalette)
  };

  return appConfigModel;
}
