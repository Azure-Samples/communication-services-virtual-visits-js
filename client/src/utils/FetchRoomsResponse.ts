// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import createError from 'http-errors';
import { CreateRoomResponse, JoinRoomResponse } from '../models/RoomModel';

export const fetchRoomsResponse = async (roomId: string, userId: string): Promise<JoinRoomResponse | undefined> => {
  const response = await fetch('/api/rooms/token', {
    method: 'POST',
    body: JSON.stringify({
      roomId: roomId,
      userId: userId
    }),
    headers: {
      'content-type': 'application/json;charset=UTF-8'
    }
  });

  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve user token. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const roomResponse = JSON.parse(responseContent);
  return roomResponse;
};

export const createRoom = async (): Promise<CreateRoomResponse> => {
  const response = await fetch('/api/rooms', {
    method: 'POST',
    headers: {
      'content-type': 'application/json;charset=UTF-8'
    }
  });
  if (response.status !== 201) {
    throw new createError(
      response.status,
      `Unable to create room. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }
  const responseContent = await response.text();
  const roomsResponse = JSON.parse(responseContent);
  return roomsResponse;
};
