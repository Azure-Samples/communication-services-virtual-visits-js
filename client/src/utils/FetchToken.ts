// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { CommunicationUserToken } from '@azure/communication-identity';
import createError from 'http-errors';

export async function fetchToken(): Promise<CommunicationUserToken | undefined> {
  const response = await fetch('/api/token', { method: 'GET' });

  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const token = JSON.parse(responseContent, parseDate);

  return token;
}

function parseDate(key: string, value: string): Date | string {
  if (key === 'expiresOn') {
    return new Date(value);
  }
  return value;
}
