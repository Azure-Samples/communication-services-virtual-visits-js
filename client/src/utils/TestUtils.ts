// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import {
  AdapterError,
  CallAdapter,
  CallAdapterState,
  CallWithChatAdapter,
  CallWithChatAdapterState
} from '@azure/communication-react';
import { act } from '@testing-library/react';

export const runFakeTimers = async (): Promise<void> => {
  await act(async () => {
    jest.useFakeTimers();
    jest.runAllTimers();
  });
};

// Inspired by ui-library unit test: https://aka.ms/AAfyg3m
export const createMockCallWithChatAdapter = (): CallWithChatAdapter => {
  const callWithChatAdapter = {} as CallWithChatAdapter;
  callWithChatAdapter.onStateChange = jest.fn();
  callWithChatAdapter.offStateChange = jest.fn();
  callWithChatAdapter.askDevicePermission = jest.fn();
  callWithChatAdapter.queryCameras = jest.fn();
  callWithChatAdapter.queryMicrophones = jest.fn();
  callWithChatAdapter.querySpeakers = jest.fn();
  callWithChatAdapter.on = jest.fn(); // allow for direct subscription to the state of the call-with-chat adapter
  callWithChatAdapter.off = jest.fn(); // Allow for direct un-subscription to the state of the call-with-chat adapter
  callWithChatAdapter.getState = jest.fn(
    (): CallWithChatAdapterState => ({
      page: 'lobby',
      isLocalPreviewMicrophoneEnabled: false,
      userId: { kind: 'communicationUser', communicationUserId: 'test' },
      displayName: 'test',
      devices: {
        isSpeakerSelectionAvailable: false,
        cameras: [],
        microphones: [],
        speakers: [],
        unparentedViews: []
      },
      isTeamsCall: true,
      call: undefined,
      chat: undefined,
      latestCallErrors: { test: new Error() as AdapterError },
      latestChatErrors: { test: new Error() as AdapterError }
    })
  );
  return callWithChatAdapter;
};

export const createMockCallAdapter = (): CallAdapter => {
  const callAdapter = {} as CallAdapter;
  callAdapter.onStateChange = jest.fn();
  callAdapter.offStateChange = jest.fn();
  callAdapter.askDevicePermission = jest.fn();
  callAdapter.queryCameras = jest.fn();
  callAdapter.queryMicrophones = jest.fn();
  callAdapter.querySpeakers = jest.fn();
  callAdapter.on = jest.fn(); // allow for direct subscription to the state of the call adapter
  callAdapter.off = jest.fn(); // Allow for direct un-subscription to the state of the call adapter
  callAdapter.getState = jest.fn(
    (): CallAdapterState => ({
      page: 'lobby',
      isLocalPreviewMicrophoneEnabled: false,
      userId: { kind: 'communicationUser', communicationUserId: 'test' },
      displayName: 'test',
      call: undefined,
      devices: {
        isSpeakerSelectionAvailable: false,
        cameras: [],
        microphones: [],
        speakers: [],
        unparentedViews: []
      },
      isTeamsCall: false,
      latestErrors: { test: new Error() as AdapterError }
    })
  );
  return callAdapter;
};

export const createMockStatefulCallClient = () => {
  return { createCallAgent: () => '' };
};

export const createMockStatefulChatClient = () => {
  return { startRealtimeNotifications: () => '', getChatThreadClient: () => '' };
};

export const createMockCallWithChatComposite = () => {
  return 'hello';
};

export const createMockCallComposite = () => {
  return 'hello';
};
