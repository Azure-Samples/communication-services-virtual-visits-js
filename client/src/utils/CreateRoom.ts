// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { RoomParticipant, RoomParticipantRole } from '../models/RoomModel';
import { createRoom } from './FetchRoomsResponse';
import { makeRoomsJoinUrl } from './GetMeetingLink';

export const createRoomAndRedirectUrl = async (): Promise<string> => {
  let redirectUrl: string;
  try {
    const roomResponse = await createRoom();
    const roomId = roomResponse.roomId;
    const userId = roomResponse.participants.find(
      (participant: RoomParticipant) => (participant.role as RoomParticipantRole) === RoomParticipantRole.presenter
    )?.id;
    if (userId) redirectUrl = makeRoomsJoinUrl(roomId, userId);
    else throw new Error(`room does not have participant with role ${RoomParticipantRole.presenter}`);
  } catch (e) {
    //EndUser to add logging errors in place of console logging below
    console.log(e);
    throw e;
  }
  return redirectUrl;
};
