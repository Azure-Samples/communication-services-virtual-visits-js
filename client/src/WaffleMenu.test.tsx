// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Panel } from '@fluentui/react';
import { mount } from 'enzyme';
import * as renderer from 'react-test-renderer';
import { WaffleMenu, WaffleNavigation } from './WaffleMenu';
import { runFakeTimers } from './utils/TestUtils';

describe('WaffleMenu', () => {
  it('should have a Waffle icon and a closed Panel', async () => {
    const waffleMenu = mount(<WaffleMenu parentid="test" />);

    const menuIcon = waffleMenu.find({ iconName: 'Waffle' });
    const panel = waffleMenu.find(Panel);

    expect(menuIcon.length).toBe(1);
    expect(panel.length).toBe(1);
    expect(panel.first().prop('isOpen')).toBe(false);
  });

  it('should open its Panel when Waffle icon is clicked on', async () => {
    const waffleMenu = mount(<WaffleMenu parentid="test" />);

    const menuIcon = waffleMenu.find({ iconName: 'Waffle' });

    menuIcon.simulate('click');

    await runFakeTimers();

    waffleMenu.update();

    const panel = waffleMenu.find(Panel);

    expect(panel.length).toBe(1);
    expect(panel.first().prop('isOpen')).toBe(true);
  });

  it('panel should have links to the book and visit page', () => {
    const waffleNavigationTree = renderer.create(<WaffleNavigation />).toJSON();
    expect(waffleNavigationTree).toMatchSnapshot();
  });
});
