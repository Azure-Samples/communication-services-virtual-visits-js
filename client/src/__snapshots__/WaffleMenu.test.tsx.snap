// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaffleMenu panel should have links to the book and visit page 1`] = `
<div
  className="ms-FocusZone css-151"
  data-focuszone-id="FocusZone8"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
>
  <nav
    aria-label="Navigate to expected page"
    className="ms-Nav root-134"
    role="navigation"
  >
    <div
      className="ms-Nav-group is-expanded"
    >
      <div
        className="ms-Nav-groupContent groupContent-142"
      >
        <ul
          className="ms-Nav-navItems navItems-141"
          role="list"
        >
          <li
            className="ms-Nav-navItem navItem-140"
            role="listitem"
          >
            <div
              className="ms-Nav-compositeLink is-selected compositeLink-136"
              name="Home"
            >
              <a
                aria-current="page"
                className="ms-Button ms-Button--action ms-Button--command ms-Nav-link link-152"
                data-is-focusable={true}
                href="./"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Home"
              >
                <span
                  className="ms-Button-flexContainer flexContainer-153"
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className="ms-Nav-linkText linkText-135"
                  >
                    Home
                  </div>
                </span>
              </a>
            </div>
          </li>
          <li
            className="ms-Nav-navItem navItem-140"
            role="listitem"
          >
            <div
              className="ms-Nav-compositeLink compositeLink-136"
              name="Book"
            >
              <a
                className="ms-Button ms-Button--action ms-Button--command ms-Nav-link link-158"
                data-is-focusable={true}
                href="./book"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Book"
              >
                <span
                  className="ms-Button-flexContainer flexContainer-153"
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className="ms-Nav-linkText linkText-135"
                  >
                    Book
                  </div>
                </span>
              </a>
            </div>
          </li>
          <li
            className="ms-Nav-navItem navItem-140"
            role="listitem"
          >
            <div
              className="ms-Nav-compositeLink compositeLink-136"
              name="Visit"
            >
              <a
                className="ms-Button ms-Button--action ms-Button--command ms-Nav-link link-158"
                data-is-focusable={true}
                href="./visit"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Visit"
              >
                <span
                  className="ms-Button-flexContainer flexContainer-153"
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className="ms-Nav-linkText linkText-135"
                  >
                    Visit
                  </div>
                </span>
              </a>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
